includes/class-sensei-frontend.php

```diff
@@ -750,35 +751,70 @@ public function lesson_tag_archive_description() {
 
 	public function sensei_complete_lesson() {
 		global $post,  $current_user;
-		// Handle Quiz Completion
-		if ( isset( $_POST['quiz_action'] ) && wp_verify_nonce( $_POST[ 'woothemes_sensei_complete_lesson_noonce' ], 'woothemes_sensei_complete_lesson_noonce' ) ) {
 
-			$sanitized_submit = esc_html( $_POST['quiz_action'] );
+		if ( ! isset( $_POST['quiz_action'] ) ) {
+			return;
+		}
 
-			switch ($sanitized_submit) {
-                case 'lesson-complete':
+		if ( ! wp_verify_nonce( $_POST[ 'woothemes_sensei_complete_lesson_noonce' ], 'woothemes_sensei_complete_lesson_noonce' ) ) {
+			return;
+		}
 
-					Sensei_Utils::sensei_start_lesson( $post->ID, $current_user->ID, $complete = true );
+		$lesson_id = intval( $_POST['lesson_id'] );
 
-					break;
+		if ( $lesson_id <= 0 ) {
+			return;
+		}
 
-                case 'lesson-reset':
+		// Handle Quiz Completion
+		$sanitized_submit = esc_html( $_POST['quiz_action'] );
 
-					Sensei_Utils::sensei_remove_user_from_lesson( $post->ID, $current_user->ID );
+		switch ($sanitized_submit) {
+              case 'lesson-complete':
+				Sensei_Utils::sensei_start_lesson( $lesson_id, $current_user->ID, $complete = true );
+				$this->maybe_redirect_to_next_lesson( $lesson_id );
 
-					$this->messages = '<div class="sensei-message note">' .  __( 'Lesson Reset Successfully.', 'woothemes-sensei' ) . '</div>';
-					break;
+				break;
 
-				default:
-					// Nothing
-					break;
+              case 'lesson-reset':
+				Sensei_Utils::sensei_remove_user_from_lesson( $lesson_id, $current_user->ID );
 
-			} // End Switch Statement
+				$this->messages = '<div class="sensei-message note">' .  __( 'Lesson Reset Successfully.', 'woothemes-sensei' ) . '</div>';
```

Comment:
There's no message delivery framework available? Repeatedly composing HTML for every message seems much porcelain.

Reply:
I'm not planning to open the messaging can of worms on this PR. 😅